src/multipart.rs

```diff
@@ -0,0 +1,172 @@
+use futures::{future::BoxFuture, stream::FuturesUnordered, Future, StreamExt};
+use std::{io, pin::Pin, sync::Arc, task::Poll};
+use tokio::io::AsyncWrite;
+
+use crate::Result;
+
+type BoxedTryFuture<T> = Pin<Box<dyn Future<Output = Result<T, io::Error>> + Send>>;
+
+// Lifetimes are difficult to manage, so not using AsyncTrait
+pub(crate) trait CloudMultiPartUploadImpl {
+    /// Upload a single part
+    fn put_multipart_part(
+        &self,
+        buf: Vec<u8>,
+        part_idx: usize,
+    ) -> BoxFuture<'static, Result<(usize, UploadPart), io::Error>>;
+
+    /// Complete the upload with the provided parts
+    fn complete(
+        &self,
+        completed_parts: Vec<Option<UploadPart>>,
+    ) -> BoxFuture<'static, Result<(), io::Error>>;
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct UploadPart {
+    pub content_id: String,
+}
+
+pub(crate) struct CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl,
+{
+    inner: Arc<T>,
+    /// A list of completed parts, in sequential order.
+    completed_parts: Vec<Option<UploadPart>>,
+    /// Part upload tasks currently running
+    tasks: FuturesUnordered<BoxedTryFuture<(usize, UploadPart)>>,
+    /// Maximum number of upload tasks to run concurrently
+    max_concurrency: usize,
+    /// Buffer that will be sent in next upload.
+    current_buffer: Vec<u8>,
+    /// Minimum size of a part in bytes
+    min_part_size: usize,
+    /// Index of current part
+    current_part_idx: usize,
+    /// The completion task
+    completion_task: Option<BoxedTryFuture<()>>,
+}
+
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl,
+{
+    pub fn new(inner: T, max_concurrency: usize) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            completed_parts: Vec::new(),
+            tasks: FuturesUnordered::new(),
+            max_concurrency,
+            current_buffer: Vec::new(),
+            // TODO: Should self vary by provider?
+            // TODO: Should we automatically increase then when part index gets large?
+            min_part_size: 5_000_000,
+            current_part_idx: 0,
+            completion_task: None,
+        }
+    }
+
+    pub fn poll_tasks(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Result<(), io::Error> {
+        if self.tasks.is_empty() {
+            return Ok(());
+        }
+        let total_parts = self.completed_parts.len();
+        while let Poll::Ready(Some(res)) = self.tasks.poll_next_unpin(cx) {
```

Comment:
Technically I think this should be checking https://docs.rs/futures/latest/futures/stream/trait.FusedStream.html#tymethod.is_terminated to ensure it isn't polling a stream that has already yielded `None`

Reply:
Looking at the source code of FuturesUnordered, I don't see any bad consequences of running `self.tasks.poll_next()` without checking if it's terminated. It should just keep returning `None`.

And it's quite possible that we will return `None` after flushing some data, then add new request futures to the queue and start polling those again.

So I'm not sure if adding such a check will add anything. But let me know if you see a flaw in my understanding.

Reply:
As I said, I think it is probably fine, but technically polling a stream after it has returned None is not well defined (at the trait level).